In [2]:
import pandas as pd
import numpy as np
pd.__version__

'1.3.1'

In [3]:
pd.options.display.float_format = '{:,.2f}'.format

In [4]:
import seaborn as sns

Cargamos una base de datos de propinas

In [26]:
df = sns.load_dataset('tips')
df

total_bill  tip     sex smoker   day    time  size
0         16.99 1.01  Female     No   Sun  Dinner     2
1         10.34 1.66    Male     No   Sun  Dinner     3
2         21.01 3.50    Male     No   Sun  Dinner     3
3         23.68 3.31    Male     No   Sun  Dinner     2
4         24.59 3.61  Female     No   Sun  Dinner     4
..          ...  ...     ...    ...   ...     ...   ...
239       29.03 5.92    Male     No   Sat  Dinner     3
240       27.18 2.00  Female    Yes   Sat  Dinner     2
241       22.67 2.00    Male    Yes   Sat  Dinner     2
242       17.82 1.75    Male     No   Sat  Dinner     2
243       18.78 3.00  Female     No  Thur  Dinner     2

[244 rows x 7 columns]

Algo que podemos ahcer para concer nuestros datos, es usar la función `describe`
Vemos que en día tenemos 4 categorías, en time tenemos 2 categorías, para ver un resumen de una de esas categorías podemos usar `value_counts`

In [ ]:
df.describe(include='all')

total_bill     tip   sex smoker  day    time    size
count      244.000 244.000   244    244  244     244 244.000
unique         nan     nan     2      2    4       2     nan
top            nan     nan  Male     No  Sat  Dinner     nan
freq           nan     nan   157    151   87     176     nan
mean        19.786   2.998   NaN    NaN  NaN     NaN   2.570
std          8.902   1.384   NaN    NaN  NaN     NaN   0.951
min          3.070   1.000   NaN    NaN  NaN     NaN   1.000
25%         13.348   2.000   NaN    NaN  NaN     NaN   2.000
50%         17.795   2.900   NaN    NaN  NaN     NaN   2.000
75%         24.127   3.562   NaN    NaN  NaN     NaN   3.000
max         50.810  10.000   NaN    NaN  NaN     NaN   6.000

In [20]:
df['day'].value_counts()

Sat     87
Sun     76
Thur    62
Fri     19
Name: day, dtype: int64

In [21]:
df['time'].value_counts()

Dinner    176
Lunch      68
Name: time, dtype: int64

Vemos que el restaurante solo abre de jueves a domingo y que ofrece 2 tiempos, comida y cena, (lunch , dinner).

podemos ver el porcentaje que corresponde por día :


In [23]:
#el 35% corresponde a los sábados
df['day'].value_counts() / df['day'].value_counts().sum() * 100

Sat    35.66
Sun    31.15
Thur   25.41
Fri     7.79
Name: day, dtype: float64

Otra forma de obtener el pòrcentaje de cada día, es usando el parámetro `normalize`. Con `dropna=False`, incluimos valores NA

In [54]:
df['day'].value_counts(normalize=True, dropna=False)*100

Sat    35.66
Sun    31.15
Thur   25.41
Fri     7.79
Name: day, dtype: float64

Podemos agregar una columna que corresponda al porcentaje de la propina con respecto al monto


In [27]:
df['porc_tip'] = df['tip']/df['total_bill']
df

total_bill  tip     sex smoker   day    time  size  porc_tip
0         16.99 1.01  Female     No   Sun  Dinner     2      0.06
1         10.34 1.66    Male     No   Sun  Dinner     3      0.16
2         21.01 3.50    Male     No   Sun  Dinner     3      0.17
3         23.68 3.31    Male     No   Sun  Dinner     2      0.14
4         24.59 3.61  Female     No   Sun  Dinner     4      0.15
..          ...  ...     ...    ...   ...     ...   ...       ...
239       29.03 5.92    Male     No   Sat  Dinner     3      0.20
240       27.18 2.00  Female    Yes   Sat  Dinner     2      0.07
241       22.67 2.00    Male    Yes   Sat  Dinner     2      0.09
242       17.82 1.75    Male     No   Sat  Dinner     2      0.10
243       18.78 3.00  Female     No  Thur  Dinner     2      0.16

[244 rows x 8 columns]

Veamos  la media por género:

En promedio los hombres gastan más que las mujeres, y a la vez su propina es mayor, sin embargo con respecto al monto, es mayor el de las mujeres: 0.17

In [31]:
df.groupby('sex').mean()

total_bill  tip  size  porc_tip
sex                                    
Male         20.74 3.09  2.63      0.16
Female       18.06 2.83  2.46      0.17

In [33]:
#vemos que la distribución está casi estandarizada
df.groupby('sex').median()

total_bill  tip  size  porc_tip
sex                                    
Male         18.35 3.00  2.00      0.15
Female       16.40 2.75  2.00      0.16

In [35]:
df.groupby('sex')[['porc_tip']].describe()

porc_tip                                   
          count mean  std  min  25%  50%  75%  max
sex                                               
Male     157.00 0.16 0.06 0.04 0.12 0.15 0.19 0.71
Female    87.00 0.17 0.05 0.06 0.14 0.16 0.19 0.42

In [36]:
df.groupby('sex')[['total_bill','porc_tip']].describe()

total_bill                                         porc_tip            \
            count  mean  std  min   25%   50%   75%   max    count mean  std   
sex                                                                            
Male       157.00 20.74 9.25 7.25 14.00 18.35 24.71 50.81   157.00 0.16 0.06   
Female      87.00 18.06 8.01 3.07 12.75 16.40 21.52 44.30    87.00 0.17 0.05   

                                 
        min  25%  50%  75%  max  
sex                              
Male   0.04 0.12 0.15 0.19 0.71  
Female 0.06 0.14 0.16 0.19 0.42

In [37]:
df.groupby('sex')[['total_bill','porc_tip']].mean()

total_bill  porc_tip
sex                         
Male         20.74      0.16
Female       18.06      0.17

In [12]:
def mean_eur2usd(x):
    y = np.mean(x)*1.12
    return y

In [13]:
mean_eur2usd([100,102])

113.12

In [38]:
df.groupby('sex')[['total_bill','porc_tip']].apply(mean_eur2usd)

total_bill  porc_tip
sex                         
Male         23.23      0.18
Female       20.22      0.19

Puedo aplicar una función a cada columna usando `apply` , pueden ser funciones definidas por nosotros o built-in

In [39]:
df.groupby(['sex','time'])[['total_bill','porc_tip']].apply(mean_eur2usd)

total_bill  porc_tip
sex    time                        
Male   Lunch        20.21      0.19
       Dinner       24.04      0.17
Female Lunch        18.30      0.18
       Dinner       21.52      0.19

In [43]:
df.groupby(['sex','time'])[['total_bill','porc_tip']].mean()

total_bill  porc_tip
sex    time                        
Male   Lunch        18.05      0.17
       Dinner       21.46      0.16
Female Lunch        16.34      0.16
       Dinner       19.21      0.17

Vemos que la variación es de alrededor de 8 dolares

In [41]:
df.groupby(['sex','time'])[['total_bill','porc_tip']].apply(np.std)

total_bill  porc_tip
sex    time                        
Male   Lunch         7.83      0.05
       Dinner        9.42      0.07
Female Lunch         7.39      0.03
       Dinner        8.12      0.06

También podemos aplicar funciones anónimas con lambda. Usando Lambda el ejemploq uedaría como:

In [44]:
df.groupby(['sex','time'])[['total_bill','porc_tip']].apply(lambda x: np.mean(x)*1.12)

total_bill  porc_tip
sex    time                        
Male   Lunch        20.21      0.19
       Dinner       24.04      0.17
Female Lunch        18.30      0.18
       Dinner       21.52      0.19

Podemos ver que en la cena los hombres tiene una cuenta mayor en promedio y además de valor maximo, y que los hombres van en grupos mayores durante la cena.
Hay algo interesante, hay un el porcentaje de propina que alcanzó el 71% del valor de la cuenta.

In [45]:
df.groupby(['sex','time'])[['total_bill','porc_tip','size']].aggregate([np.mean,np.max])

total_bill       porc_tip      size     
                    mean  amax     mean amax mean amax
sex    time                                           
Male   Lunch       18.05 41.19     0.17 0.27 2.36    6
       Dinner      21.46 50.81     0.16 0.71 2.70    6
Female Lunch       16.34 43.11     0.16 0.26 2.46    6
       Dinner      19.21 44.30     0.17 0.42 2.46    5

In [50]:
df.groupby(['sex','time'])[['total_bill','porc_tip','size']].aggregate(['mean',max])

total_bill       porc_tip      size    
                    mean   max     mean  max mean max
sex    time                                          
Male   Lunch       18.05 41.19     0.17 0.27 2.36   6
       Dinner      21.46 50.81     0.16 0.71 2.70   6
Female Lunch       16.34 43.11     0.16 0.26 2.46   6
       Dinner      19.21 44.30     0.17 0.42 2.46   5

In [ ]:
df.groupby(['sex','time'])[['total_bill','prct_tip']].agg([np.mean,np.max])

total_bill        prct_tip      
                    mean   amax     mean  amax
sex    time                                   
Male   Lunch      18.048 41.190    0.166 0.266
       Dinner     21.461 50.810    0.155 0.710
Female Lunch      16.339 43.110    0.162 0.259
       Dinner     19.213 44.300    0.169 0.417

In [55]:
dict_agg = {'tip':[min, max],'total_bill':[np.mean, mean_eur2usd]}
dict_agg

{'tip': [<function min>, <function max>],
 'total_bill': [<function numpy.mean(a, axis=None, dtype=None, out=None, keepdims=<no value>, *, where=<no value>)>,
  <function __main__.mean_eur2usd(x)>]}

In [58]:
df.groupby(['sex','time'])[['total_bill','tip']].agg(dict_agg)

tip       total_bill             
               min   max       mean mean_eur2usd
sex    time                                     
Male   Lunch  1.44  6.70      18.05        20.21
       Dinner 1.00 10.00      21.46        24.04
Female Lunch  1.25  5.17      16.34        18.30
       Dinner 1.00  6.50      19.21        21.52

No es necesario especificar las columnas `[['total_bill','tip']]` , pues ya están especificadas en el dictionario

In [59]:
df.groupby(['sex','time']).agg(dict_agg)

tip       total_bill             
               min   max       mean mean_eur2usd
sex    time                                     
Male   Lunch  1.44  6.70      18.05        20.21
       Dinner 1.00 10.00      21.46        24.04
Female Lunch  1.25  5.17      16.34        18.30
       Dinner 1.00  6.50      19.21        21.52

Vemos que en el resultado anterior la propina másxima fue de 10 Euros, podemosintentar filtrar 



In [60]:
#donde el total de la cuenta promedio sea de mas de 20 dolares
def f_filter(x):
    return mean_eur2usd(x['total_bill'].mean()) > 20

Lo que hace es filtrar por grupo de género y tiempo, las cuentas que en promedio sean de mayor de 20 dolares, que son todas menos la comida de las mujeres, pies el valor medio de éstas es de 18.3 $

In [61]:
df_filtered = df.groupby(['sex','time']).filter(f_filter)
df_filtered

total_bill  tip     sex smoker   day    time  size  porc_tip
0         16.99 1.01  Female     No   Sun  Dinner     2      0.06
1         10.34 1.66    Male     No   Sun  Dinner     3      0.16
2         21.01 3.50    Male     No   Sun  Dinner     3      0.17
3         23.68 3.31    Male     No   Sun  Dinner     2      0.14
4         24.59 3.61  Female     No   Sun  Dinner     4      0.15
..          ...  ...     ...    ...   ...     ...   ...       ...
239       29.03 5.92    Male     No   Sat  Dinner     3      0.20
240       27.18 2.00  Female    Yes   Sat  Dinner     2      0.07
241       22.67 2.00    Male    Yes   Sat  Dinner     2      0.09
242       17.82 1.75    Male     No   Sat  Dinner     2      0.10
243       18.78 3.00  Female     No  Thur  Dinner     2      0.16

[209 rows x 8 columns]

In [62]:
df_filtered.groupby(['sex','time']).count()

total_bill  tip  smoker  day  size  porc_tip
sex    time                                                
Male   Lunch           33   33      33   33    33        33
       Dinner         124  124     124  124   124       124
Female Lunch            0    0       0    0     0         0
       Dinner          52   52      52   52    52        52